### PR TITLE
[#103] [Backend] Add Get Comments API

### LINKED ISSUE #103: As a user, I can see the list of comments of an article
## Why
The users will be able to see the list of comments on any article from the server with optional authentication.

## Acceptance Criteria
- [ ] Implement the API for getting the list of comments on an article via this [endpoint](https://github.com/gothinkster/realworld/tree/master/api#get-comments-from-an-article).
- [ ] The endpoint takes the article's `slug` as the parameter in the request url.
- [ ] Create a usecase for getting the list of comments on an article for later usage in integrate task.
- [ ] Parse the response into the corresponding model, here is an example response for getting the list of comments on an article:
```
{
  "comments": [{
    "id": 1,
    "createdAt": "2016-02-18T03:22:56.637Z",
    "updatedAt": "2016-02-18T03:22:56.637Z",
    "body": "It takes a Jacobian",
    "author": {
      "username": "jake",
      "bio": "I work at statefarm",
      "image": "https://i.stack.imgur.com/xHWG8.jpg",
      "following": false
    }
  }]
}
```

### PR DESCRIPTION
**Pull request by markgravity**
Resolved #103

## What happened

Add Get Comments API

## Insight

- [x] Implement the API for getting the list of comments on an article via this [endpoint](https://github.com/gothinkster/realworld/tree/master/api#get-comments-from-an-article).
- [x] The endpoint takes the article's `slug` as the parameter in the request url.
- [x] Create a usecase for getting the list of comments on an article for later usage in integrate task.
- [x] Parse the response into the corresponding model

## Proof Of Work

Request Log
```
*** Overview *** 
URL https://conduit.productionready.io/api/articles/article-created-by-cypress-test-12gz5t/comments
Method GET
Response Code 200
Request Start Time Sep 15 2021 - 14:02:02
Duration 1s

*** Request Header *** 
Accept-Language en;q=1.0
Accept-Encoding br;q=1.0, gzip;q=0.9, deflate;q=0.8
User-Agent Nimble Medium - Staging/1.0.1 (co.nimblehq.nimble-medium-stag; build:1; iOS 14.5.0) Alamofire/5.4.3


*** Request Body *** 
-

*** Response Header *** 
Alt-Svc h3=":443"; ma=86400, h3-29=":443"; ma=86400, h3-28=":443"; ma=86400, h3-27=":443"; ma=86400
x-runtime 0.061863
nel {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-powered-by Phusion Passenger 4.0.60
Etag W/"183b0d8cbd0b9e5a4e8a0dcf3b8bec4b"
Server cloudflare
cf-cache-status DYNAMIC
Vary Origin
x-xss-protection 1; mode=block
x-request-id f6dd68b9-8476-4bf2-8cf0-58764af18dbe
expect-ct max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
x-content-type-options nosniff
Cache-Control max-age=0, private, must-revalidate
cf-ray 68effcc2dd564c95-SIN
report-to {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=LJhSKh6FCj0OMdnolnetImbcHmzfZJdtBEBNycvPgej0GmGQ6tmXLfGPapa5nI7fuHWlzme%2Frn2rLgat3qIU0UNmktwiznbl3qFCXKygyhoKduiOj9prOOOvpUdzWIQPGE6t5y1agdomefjrbg%3D%3D"}],"group":"cf-nel","max_age":604800}
Content-Encoding br
x-frame-options SAMEORIGIN
Status 200 OK
Date Wed, 15 Sep 2021 07:02:04 GMT
Content-Type application/json; charset=utf-8


*** Response Body *** 
{
  "comments" : [
    {
      "body" : "Cypress comment",
      "id" : 106061,
      "author" : {
        "username" : "cy831845063",
        "image" : "https://static.productionready.io/images/smiley-cyrus.jpg",
        "following" : false,
        "bio" : null
      },
      "updatedAt" : "2021-09-15T06:06:05.129Z",
      "createdAt" : "2021-09-15T06:06:05.129Z"
    }
  ]
}

------------------------------------------------------------------------
------------------------------------------------------------------------
------------------------------------------------------------------------
```

Decode Log
```
2021-09-15 14:02:02.707: getComments -> subscribed
2021-09-15 14:02:05.780: getComments -> Event next([NimbleMedium.APIArticleComment(id: 106061, createdAt: 2021-09-15 06:06:05 +0000, updatedAt: 2021-09-15 06:06:05 +0000, body: "Cypress comment", apiAuthor: NimbleMedium.DecodableProfile(username: "cy831845063", bio: nil, image: Optional("https://static.productionready.io/images/smiley-cyrus.jpg"), _following: DefaultCodable.Default<DefaultCodable.False>(wrappedValue: false)))])
2021-09-15 14:02:05.780: getComments -> Event completed
2021-09-15 14:02:05.780: getComments -> isDisposed
```

### CODE REVIEW

File: NimbleMedium/Sources/Injection/Resolver+Injection.swift [hunk 1]
```diff
@@ -26,6 +27,9 @@ extension Resolver: ResolverRegistering {
         register {
             ArticleRepository(networkAPI: resolve())
         }.implements(ArticleRepositoryProtocol.self)
+        register {
+            ArticleCommentRepository(networkAPI: resolve())
+        }.implements(ArticleCommentRepositoryProtocol.self)
 
         // UseCases
         register {
```
Comment (minhnimble) [inline NimbleMedium/Sources/Injection/Resolver+Injection.swift:35]: maybe we can refactor this place bit to create 3 separate functions for registering UseCases, Repositories and ViewModels and including it in here, preventing a single long function and fix the lint issue, what do you think? 🙏

```
public static func registerAllServices() {
...
registerRepositories()
registerUsecases()
registerViewModels()
}

private static func registerRepositories() {
  // Register Repositories code here
  ...
}

// Same for UserCases and ViewModels 
...
```

File: NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleCommentRepositorySpec.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    let inputResponse = APIArticleCommentsResponse.dummy
+
+                    beforeEach {
+                        outputArticleComments = scheduler.createObserver([APIArticleComment].self)
+                        networkAPI.setPerformRequestForReturnValue(Single.just(inputResponse))
+                        repository.getComments(slug: "")
+                        .asObservable()
+                        .map {
+                            $0.compactMap { $0 as? APIArticleComment }
+                        }
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleCommentRepositorySpec.swift:47]: can one line this I think, it doesn't exceed the max line length constraint yet 😁
```suggestion
                        .map { $0.compactMap { $0 as? APIArticleComment } }
```

File: NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleCommentRepositorySpec.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                }
+
+                context("when the request returns failure") {
+
+                    var outputError: TestableObserver<Error?>!
+
+                    beforeEach {
+                        outputError = scheduler.createObserver(Optional<Error>.self)
+                        // swiftlint:disable line_length
+                        networkAPI.setPerformRequestForReturnValue(Single<APIArticleCommentsResponse>.error(TestError.mock))
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleCommentRepositorySpec.swift:64]: maybe we should update to use multiple lines for here and remove the lint check disabling 🙏

File: NimbleMediumTests/Sources/Specs/Domain/UseCases/GetArticleCommentsUseCaseSpec.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+                    beforeEach {
+                        outputArticleComments = scheduler.createObserver([APIArticleComment].self)
+                        articleCommentRepository.getCommentsSlugReturnValue = .just(inputArticleComments)
+
+                        usecase.getComments(slug: "")
+                        .asObservable()
+                        .map {
+                            $0.compactMap { $0 as? APIArticleComment }
+                        }
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Domain/UseCases/GetArticleCommentsUseCaseSpec.swift:48]: same here
```suggestion
                        .map { $0.compactMap { $0 as? APIArticleComment } }
```

**Review by minhnimble — APPROVED** [review]

**Review by suho — APPROVED** [review]

**Comment by minhnimble** [comment]
LGTM 🚀 Just a friendly reminder that please only merge it after the father branch of this PR is merged first 🙏

**Pull request merged**